### PR TITLE
THREESCALE-11281 - bump to go 1.22

### PR DESCRIPTION
**Pull request by valerymo**
Jira: https://issues.redhat.com/browse/THREESCALE-11281

## Verification 

1. Install and sanity validation
3scale installed locally

```
~/go/3scale-operator git branch --show-current
THREESCALE-11281-1


cd 3scale-operator
make install
export NAMESPACE=3scale-test
oc new-project $NAMESPACE
oc project $NAMESPACE
make download
make run
```
Installed correctly
routes/UI verified

```
~/go/3scale-operator oc project
Using project "3scale-test" on server "https://api.vmo01.xxxxx".
~/go/3scale-operator date                     
Wed Nov 27 19:17:24 EET 2024
~/go/3scale-operator git branch --show-current
THREESCALE-11281-1
~/go/3scale-operator oc get deploy            
NAME                 READY   UP-TO-DATE   AVAILABLE   AGE
apicast-production   1/1     1            1           36m
apicast-staging      1/1     1            1           36m
backend-cron         1/1     1            1           40m
backend-listener     1/1     1            1           40m
backend-redis        1/1     1            1           40m
backend-worker       1/1     1            1           40m
system-app           1/1     1            1           34m
system-memcache      1/1     1            1           40m
system-mysql         1/1     1            1           40m
system-redis         1/1     1            1           40m
system-searchd       1/1     1            1           40m
system-sidekiq       1/1     1            1           36m
zync                 1/1     1            1           36m
zync-database        1/1     1            1           36m
zync-que             1/1     1            1           36m
~/go/3scale-operator 


```

2. CVEs analysys
Quay.io Security Scan - Passed

- Image was built to see CVE check in Quay.io.
- Repo: quay.io/vmogilev_rhmi/3scale-operator:test-go122
- Command used to build image (on Mac arm):
```
docker build . --platform linux/amd64 -t quay.io/vmogilev_rhmi/3scale-operator:test-go122
```

image

image

### CODE REVIEW

File: PROJECT [hunk 1]
```diff
@@ -1,51 +1,132 @@
...
-layout: go.kubebuilder.io/v2
+layout: go.kubebuilder.io/v3
 multigroup: true
 projectName: 3scale-operator
 repo: github.com/3scale/3scale-operator
 resources:
-- group: apps
+-
+  # TODO(user): Uncomment the below line if this resource implements a controller, else delete it.
+  # controller: true
```
Comment (MStokluska) [inline PROJECT:9]: This should be uncommented right?

File: PROJECT [hunk 2]
```diff
@@ -1,51 +1,132 @@
...
+  group: apps
   kind: APIManager
+  # TODO(user): Update the package path for your API if the below value is incorrect.
+  path: k8s.io/api/apps/v1alpha1
   version: v1alpha1
-- group: apps
+-
+  # TODO(user): Uncomment the below line if this resource implements a controller, else delete it.
+  # controller: true
```
Comment (MStokluska) [inline PROJECT:17]: This too? Can you review this file again please

File: PROJECT [hunk 3]
```diff
@@ -1,51 +1,132 @@
...
 repo: github.com/3scale/3scale-operator
 resources:
-- group: apps
+-
+  # TODO(user): Uncomment the below line if this resource implements a controller, else delete it.
+  # controller: true
+  group: apps
   kind: APIManager
+  # TODO(user): Update the package path for your API if the below value is incorrect.
+  path: k8s.io/api/apps/v1alpha1
```
Comment (MStokluska) [inline PROJECT:13]: this is wrong

File: PROJECT (hunk 1 above)
Comment (valerymo) [reply PROJECT]: done. Thank you

File: PROJECT (hunk 2 above)
Comment (valerymo) [reply PROJECT]: done. Thank you

File: PROJECT (hunk 3 above)
Comment (valerymo) [reply PROJECT]: fixed. Thank you

File: Makefile [hunk 4]
```diff
@@ -281,6 +286,9 @@ bundle-validate: $(OPERATOR_SDK)
 
 .PHONY: bundle-update-test
 bundle-update-test:
+	# Remove the createdAt field before checking the diff
+    sed -i '/createdAt:/d' ./bundle/manifests/3scale-operator.clusterserviceversion.yaml
```
[the remark this answers is not in the document]
Comment (MStokluska) [reply Makefile:290]: you can remove the createdAt from the config/manifest/bases/3scale-opreator.clusterserviceversion.yaml
The flag is updated if present, if not, it's not.

File: bundle/manifests/3scale-operator.clusterserviceversion.yaml [hunk 5]
```diff
@@ -1161,4 +1166,25 @@ spec:
   maturity: stable
   provider:
     name: Red Hat
+  relatedImages:
```
Comment (MStokluska) [inline bundle/manifests/3scale-operator.clusterserviceversion.yaml:1169]: would be nice to:
- figure out where this is coming from / why?
- redis and psql have no "name" values - why
- check midstream for possible clash when updating these values for prod release.

File: bundle/manifests/3scale-operator.clusterserviceversion.yaml [hunk 6]
```diff
@@ -240,12 +240,13 @@ metadata:
     categories: Integration & Delivery
     certified: "false"
     containerImage: quay.io/3scale/3scale-operator:master
-    createdAt: "2019-05-30T22:40:00Z"
+    createdAt: "2024-11-27T13:51:25Z"
```
Comment (MStokluska) [inline bundle/manifests/3scale-operator.clusterserviceversion.yaml:243]: as above, if we remove it from base we won't have it, I don't think this is a mandatory field for csv validation. Please remove

**Comment by valerymo** [comment]
> @valerymo left some comments, in general, could you explain to me why are we pinning operator-sdk bump to go bump? If operator SDK bump is really required, we should follow (or at least review) the migration steps in go operators here: https://sdk.operatorframework.io/docs/upgrading-sdk-version/v1.28.0/ - that said, I think stuff like "controllers should now be in internal/controllers" could potentially be skipped, but it would be good to review the migration required from 1.2 to 1.28 just to be extra cautious...




@MStokluska , I had problems using  go 1.22 with operator-sdk 1.2. So the question was - which verion of operator-sdk to migrage. Here’s a summary of the versions and their compatibility with Go 1.22.

| **Go Version** | **Operator SDK Version** | **Notes** |
|----------------|--------------------------|-----------|
| Go 1.18 - 1.20 | Operator SDK 1.2         | Works well with Go 1.18 and 1.19, but may cause issues with Go 1.22 due to Go modules changes. |
| Go 1.21 - 1.22 | Operator SDK 1.28+       | Fully compatible with Go 1.21 and 1.22. Recommended version for Go 1.22. |

After using perator-sdk 1.28 - all builds and tests worked fine.

File: bundle/manifests/apps.3scale.net_apimanagerrestores.yaml [hunk 7]
```diff
@@ -2,7 +2,7 @@ apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.9.2
+    controller-gen.kubebuilder.io/version: v0.14.0
```
[the remark this answers is not in the document]
Comment (valerymo) [reply bundle/manifests/apps.3scale.net_apimanagerrestores.yaml:5]: controller-gen  v0.9.2 make a problem with GO 1.22 - following step failed: assets-validate, bundle-validate, run-unit-tests, test-crds. All with same/similar error in controller-gen  - segmentation fault , as for example
```
/home/circleci/project/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xa00d6f]
```
Looks like known Issue -  https://github.com/kubernetes-sigs/controller-tools/issues/880 . People using newer controller-gen 0.14 . Same version used also in marin3r.

**Comment by valerymo** [comment]
This PR will be closed, it's replaced by new one https://github.com/3scale/3scale-operator/pull/1045 - that does not change OperatorSDK. 
- We have verified that 3scale-operator with the existing version of OperatorSDK 1.2 works with GO 1.22.
- controller-gen update was done same way as here,  from v0.9.2 to v0.14.0 due to incompatibility of controller-gen v0.9.2 with GO 1.22 (segmentation fault).
- The SDK upgrade will be addressed in a future task. cc @MStokluska @carlkyrillos

**Pull request closed without merging**